minecraft/src/main/java/net/fabricmc/loader/impl/game/minecraft/BundlerProcessor.java

```diff
@@ -0,0 +1,118 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.loader.impl.game.minecraft;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+
+import net.fabricmc.api.EnvType;
+import net.fabricmc.loader.impl.game.minecraft.LibClassifier.Lib;
+import net.fabricmc.loader.impl.util.LoaderUtil;
+
+final class BundlerProcessor {
+	private static final String MAIN_CLASS_PROPERTY = "bundlerMainClass";
+
+	static void process(LibClassifier classifier) throws IOException {
+		Path bundlerOrigin = classifier.getOrigin(Lib.MC_BUNDLER);
+
+		// determine urls by running the bundler and extracting them from the context class loader
+
+		URL[] urls;
+
+		try (URLClassLoader bundlerCl = new URLClassLoader(new URL[] { bundlerOrigin.toUri().toURL() }, MinecraftGameProvider.class.getClassLoader()) {
+			@Override
+			protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+				synchronized (getClassLoadingLock(name)) {
+					Class<?> c = findLoadedClass(name);
+
+					if (c == null) {
+						if (name.startsWith("net.minecraft.")) {
+							URL url = getResource(LoaderUtil.getClassFileName(name));
+
+							if (url != null) {
+								try (InputStream is = url.openConnection().getInputStream()) {
+									byte[] data = new byte[Math.max(is.available() + 1, 1000)];
+									int offset = 0;
+									int len;
+
+									while ((len = is.read(data, offset, data.length - offset)) >= 0) {
+										offset += len;
+										if (offset == data.length) data = Arrays.copyOf(data, data.length * 2);
+									}
+
+									c = defineClass(name, data, 0, offset);
+								} catch (IOException e) {
+									throw new RuntimeException(e);
+								}
+							}
+						}
+
+						if (c == null) {
+							c = getParent().loadClass(name);
+						}
+					}
+
+					if (resolve) {
+						resolveClass(c);
+					}
+
+					return c;
+				}
+			}
+		}) {
+			Class<?> cls = Class.forName(classifier.getClassName(Lib.MC_BUNDLER), true, bundlerCl);
+			Method method = cls.getMethod("main", String[].class);
+
+			// save + restore the system property and context class loader just in case
+
+			String prevProperty = System.getProperty(MAIN_CLASS_PROPERTY);
+			System.setProperty(MAIN_CLASS_PROPERTY, BundlerClassPathCapture.class.getName());
+
+			ClassLoader prevCl = Thread.currentThread().getContextClassLoader();
+			Thread.currentThread().setContextClassLoader(bundlerCl);
+
+			method.invoke(null, (Object) new String[0]);
+			urls = BundlerClassPathCapture.FUTURE.get(10, TimeUnit.SECONDS);
+
+			Thread.currentThread().setContextClassLoader(prevCl);
```

Comment:
Should these restoration code be put into a `finally` clause in case the main method invocation or the future execution throws anything?

Reply:
good point, the ClassNotFoundException path may not be restored fully otherwise.